Read/write records and features to/from local db instead of remote

gino-m opened the pull request:
Closes #19.

For velocity's sake, I'll keep adding commits to this PR until you're ready to take a look. There are many dependent changes that are difficult to split into separate PRs. Thanks!!

scolsen commented on gnd/src/main/java/com/google/android/gnd/persistence/local/LocalDataStore.java:
```diff
@@ -28,8 +37,27 @@
+   * Returns a long-lived stream that emits the full set of features for a project on subscribe, and
+   * continues to return the full set each time a feature is added/changed/removed. The full set is
+   * returned rather than deltas for simplicity and to implement a fully reactive UI in which each
+   * update is idempotent.
```
Sorry for the lengthy comment! 

I understand the first part of this comment--I can see how returning the full set is simpler than returning deltas--but I don't fully understand the second part of this comment. It seems like the UI layer is actually *less* reactive by operating on the entire set? To me, reactivity in this case would suggest that the UI reacts to changes to the underlying data structure, in this case the set, i.e. the UI is set up to "react" to alterations to the set, the flowable would simply propagate these changes to the UI. 

I'm wondering if this stream is at the right level of granularity. In general I'm suspicious of streams that emit iterable data structures for reasons other than semantic correspondence. For example, a stream of cards in a deck is different semantically from a stream of decks of cards where these decks differ, e.g. by language. That seems like a valid distinction and a valid place to use a stream that emits something that's iterable (decks) because we care about some property that emerges at the level of the deck (language). 

I think this case is actually one of the former--we are really dealing with cards (features) here, we care about differences at the card level and we are only using a single deck, not comparing decks (e.g. when do we add a new card into a single deck, when do we remove a card from the deck, when do we change the face of a card in the deck), not at the deck level (when do we get a new deck, when do we change the language of an entire deck, when do we remove an entire deck--in our terms this would be something akin to caring about when *all* features are simultaneously removed from a project, when all features are simultaneously updated etc.).

If I understand this correctly, as this is set up we will either wind up re-rendering a lot of features already rendered, or have to execute a set operation (complement) to render only those features that changed. Instead of reacting to changes on the set--I think cases like this under-utilize RX...and I think we're mostly only nudged in this direction by the design of Live Data (at least on the UI side). One of the major benefits of RX is turning structures we'd normally explicitly iterate over into declaratively established dependencies between streams and functions. We lose this benefit using the approach here, since we wind up having to iterate over the stream's contents if we don't want to do unnecessary work. (obviously, this unnecessary work is negligible for projects with few features, which makes this a non-issue for the majority of cases. But what happens if I have a project (assuming this is possible) with 10,000 features? Suddenly it seems like only rendering those features that actually changed would be much more efficient, which pushes us into iterating over the set, when we could just let RX handle this and react to feature-level emissions).

All that said, idempotency is typically a good thing, and I'm not privy to the other considerations that went into this approach that might make my musings above foolish.

gino-m replied:
Thanks for the comment! I while back I had a great conversation with @yigit about this which changed our thinking and resulted in the current implementation. 

My understanding was that to ensure LiveData works even when the UI misses some emissions, and that all subscribing UIs are consistent, LiveData imposes that UIs are state-driven. In other words, emissions in a stream represent the latest state (e.g., here's the current list of things to show), not an event to be applied to existing state (add/remove/update such and such a thing in your list).

To properly implement the event/action/delta based model, you'd need to maintain state at some point in order to have something for events to mutation. Given Android Lifecycles, it's very hard to guarantee integrity and consistency of cached states. Also, assuming the stream always emits the latest state (e.g. list of pins being show) makes it a lot easier to reason about the current state of the UI.

Although intuitively sending the full set of UI elements on each change seems inefficient, in practice the diffing done at view level turns out the work quite well. In rare cases where it's too costly, workarounds can be found.

@yigit, please cmiiw!

gino-m commented:
All uses of remote db for Feature and Record persistence have been replaced, and as such, related remote data store methods are currently unused:

![image](https://user-images.githubusercontent.com/228050/59650369-bd8da300-9153-11e9-8ed5-d5f6bf4c3b1d.png)

Once merged, we can close #19.

gino-m commented on gnd/src/main/java/com/google/android/gnd/persistence/local/room/RecordMutationEntity.java:
```diff
@@ -0,0 +1,113 @@
+   * JSON object with the new value of modified attributes after this mutation. For all other
+   * mutation types this will be null.
+   *
+   * <p>Null values in the responses indicates a response was removed/cleared.
```
Ah, good catch! One way around this would be to return an empty `JSONObject` for `DELETE` mutations, but this is semantically inconsistent, since following the same rules it would be interpreted as "nothing changed".

Representing removed responses separately or using a special placeholder is also an option, but trades one kind of complexity for another. 

For now I've tried to clarify the Javadoc; please resolve if you think it works.

scolsen replied:
I think the modified doc is great!

We can rethink the technical approach if it proves problematic in the future (in other words, if people don't read the doc string).

gino-m commented on gnd/src/main/java/com/google/android/gnd/persistence/local/room/RecordDao.java:
```diff
@@ -0,0 +1,52 @@
+   * (i.e., returns on records with state = State.DEFAULT (1)).
+   */
+  @Query("SELECT * FROM record WHERE feature_id = :featureId AND state = 1")
+  Single<List<RecordEntity>> getRecordsByFeatureId(String featureId);
```
I didn't in the article that this applies to `Flowable<List>`; I tried this in practice (i.e., before after adding a feature adding records) and it didn't fail. 

Also, https://codinginfinite.com/android-room-persistent-rxjava/ recommends using `Flowable<List>` as a workaround the the problem you're describing, so I think we're ok!

scolsen replied:
Oh, interesting! I would have expected `Flowable` queries to have the same behavior irrespective of their type parameters, but I suppose this special behavior actually makes sense--some magic is probably happening to ensure Room returns an empty list if no records exist, which would likewise ensure this Flowable never calls `onError` (but it could call `onNext` with an empty list).

gino-m commented on gnd/src/main/java/com/google/android/gnd/persistence/local/room/RoomLocalDataStore.java:
```diff
@@ -54,6 +70,42 @@ public Completable applyAndEnqueue(FeatureMutation mutation) {
+  }
+
+  @Override
+  public Single<ImmutableList<Record>> getRecords(Feature feature) {
```
Nope, we just want to load them once for now. The UI doesn't auto-update, it shows a static list that can be reloaded by going back and reopening. We can eventually add drag to reload and/or reactive model.

gino-m commented on gnd/src/main/java/com/google/android/gnd/persistence/shared/RecordMutation.java:
```diff
@@ -0,0 +1,60 @@
+import com.google.common.collect.ImmutableMap;
+import java8.util.Optional;
+
+/** Represents mutation of a record in the local to be queued for sync with remote store. */
```
Updated docs for this method and FeatureMutation. Lmkwyt!

gino-m commented on gnd/src/main/java/com/google/android/gnd/persistence/shared/RecordMutation.java:
```diff
@@ -0,0 +1,60 @@
+  public abstract String getFormId();
+
+  /**
+   * Returns a map keyed by response element id. The presence of a map entry indicates a specific
```
I agree, it's generally more robust to have semantics be represented by strongly typed data structured. In this case though, we're not exposing this implementation detail outside of the `persistence` package; the code that reads it and converts it to the shared objects is collocated with the model objects in that package. Given this it's hard to justify the carrying cost of the additional code.

Either way, for expediency can we go with the simpler model for now, and if you feel strongly file a Code Health bug to change it?

scolsen replied:
Sounds good! I won't feel particularly strongly about it until I (or others) forget what the empty values signify and open non-issue issues about "weird empty values in modified responses". It's possible that will never happen, or happen infrequently enough that it isn't an issue.

scolsen commented on gnd/src/main/java/com/google/android/gnd/persistence/local/room/RecordMutationEntity.java:
```diff
@@ -62,11 +62,11 @@
-   * mutation types this will be null.
+   * For mutations of type {@link MutationEntityType#CREATE} and {@link MutationEntityType#UPDATE},
+   * returns a {@link JSONObject} with the new values of modified form responses, with {@code null}
+   * values representing responses that were removed/cleared.
```
Awesome, edit, thanks!

gino-m commented on gnd/src/main/java/com/google/android/gnd/ui/editrecord/EditRecordViewModel.java:
```diff
@@ -88,53 +87,77 @@
+                        .doOnError(this::onSaveRecordError)
+                        // Prevent from breaking upstream.
+                        .onErrorResumeNext(Observable.never())
+                        .subscribeOn(Schedulers.io()))
```
We were supposed to do it w/Firebase as well, but in that case it just printed warnings, whereas w/Room it outright 💣💣💣s. Better to get it right now while updating the streams anyway.

gino-m commented on gnd/src/main/java/com/google/android/gnd/ui/home/featuresheet/RecordListViewModel.java:
```diff
@@ -84,12 +87,14 @@ public void loadRecordSummaries(Feature feature, Form form) {
   }
 
-  private void onFetchRecordSummariesError(Throwable t) {
+  private Single<List<Record>> onFetchRecordSummariesError(Throwable t) {
```
I think it depends on the use case. If we want to screen to load but to show an error, we may want to return an empty stream or object, whereas where we want no change to happen, we'd return a Single that never resolves as we do in other places.

gino-m commented:
I've addressed all comments. I see you've already approved, but I'll wait till tomorrow to Merge to give you a chance to address my replies if necessary. Thanks a lot for the review!

gino-m commented:
If it becomes a pain point let's revisit. Keep in mind up to G4G this is
still a prototype/demo. We should try to follow sound software design
principles in the process, but we'll really need to implement unit tests
before we can get out of experimental mode and into thinking long-term
productionization.

On Wed, Jun 19, 2019 at 11:34 AM Scott Olsen <notifications@github.com>
wrote:

> *@scolsen* commented on this pull request.
> ------------------------------
>
> In
> gnd/src/main/java/com/google/android/gnd/persistence/shared/RecordMutation.java
> <https://github.com/google/ground-android/pull/129#discussion_r295365358>:
>
> > +
> +/** Represents mutation of a record in the local to be queued for sync with remote store. */
> +@AutoValue
> +public abstract class RecordMutation extends Mutation {
> +
> +  /** Returns the UUID of the record being modified. */
> +  public abstract String getRecordId();
> +
> +  /** Returns the UUID of the feature to which the mutated record belongs. */
> +  public abstract String getFeatureId();
> +
> +  /** Returns the UUID of the form associated with this record. */
> +  public abstract String getFormId();
> +
> +  /**
> +   * Returns a map keyed by response element id. The presence of a map entry indicates a specific
>
> Sounds good! I won't feel particularly strongly about it until I (or
> others) forget what the empty values signify and open non-issue issues
> about "weird empty values in modified responses". It's possible that will
> never happen, or happen infrequently enough that it isn't an issue.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/ground-android/pull/129?email_source=notifications&email_token=AABXVUTHGN4S6V7YA4LFA5LP3JGXDA5CNFSM4HYP6DGKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB4A7ROY#discussion_r295365358>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABXVURN24OYYUQXF7W267LP3JGXDANCNFSM4HYP6DGA>
> .
>


-- 

Gino Miceli |  Developer Programs Engineer |  gmiceli@google.com |  Google
Earth Outreach Developer Relations

gino-m commented:
Sgtm! Thanks for the helpful feedback!

On Wed, Jun 19, 2019 at 11:45 AM Scott Olsen <notifications@github.com>
wrote:

> *@scolsen* commented on this pull request.
> ------------------------------
>
> In
> gnd/src/main/java/com/google/android/gnd/persistence/local/room/RecordMutationEntity.java
> <https://github.com/google/ground-android/pull/129#discussion_r295371439>:
>
> > +  @CopyAnnotations
> +  @NonNull
> +  @ColumnInfo(name = "record_id")
> +  public abstract String getRecordId();
> +
> +  @CopyAnnotations
> +  @NonNull
> +  @ColumnInfo(name = "type")
> +  public abstract MutationEntityType getType();
> +
> +  /**
> +   * For edits of type {@link MutationEntityType#CREATE} and {@link MutationEntityType#UPDATE}, a
> +   * JSON object with the new value of modified attributes after this mutation. For all other
> +   * mutation types this will be null.
> +   *
> +   * <p>Null values in the responses indicates a response was removed/cleared.
>
> I think the modified doc is great!
>
> We can rethink the technical approach if it proves problematic in the
> future (in other words, if people don't read the doc string).
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/ground-android/pull/129?email_source=notifications&email_token=AABXVUQVET4OQG7R25SZNCTP3JIDLA5CNFSM4HYP6DGKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB4BBO4Y#discussion_r295371439>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABXVUULIVAEJ54TZHHBNIDP3JIDLANCNFSM4HYP6DGA>
> .
>


-- 

Gino Miceli |  Developer Programs Engineer |  gmiceli@google.com |  Google
Earth Outreach Developer Relations
